src/alloc/isolated_alloc.rs

```diff
@@ -183,13 +192,28 @@ impl IsolatedAlloc {
     /// Allocates in multiples of one page on the host system.
     ///
     /// SAFETY: Same as `alloc()`.
-    unsafe fn alloc_huge(&mut self, layout: Layout, zeroed: bool) -> *mut u8 {
-        let layout = IsolatedAlloc::huge_normalized_layout(layout, self.page_size);
-        // SAFETY: Upheld by caller
-        let ret =
-            unsafe { if zeroed { alloc::alloc_zeroed(layout) } else { alloc::alloc(layout) } };
+    unsafe fn alloc_huge(&mut self, layout: Layout) -> *mut u8 {
+        let size = self.huge_normalized_layout(layout, self.page_size);
+        // SAFETY: Layout size being a multiple of the system pagesize is enforced
+        // by huge_normalized_layout
+        let ret = unsafe {
+            libc::mmap(
+                std::ptr::null_mut(),
+                size,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_PRIVATE | libc::MAP_ANONYMOUS,
+                -1,
+                0,
+            )
+            .cast::<u8>()
+        };
+        assert_ne!(ret.addr(), usize::MAX);
+        // Align the pointer as requested
+        let ofs = ret.addr().next_multiple_of(layout.align()).strict_sub(ret.addr());
```

Comment:
Instead of this "sub" with a later "add", you could do something like `ret.map_addr(|a| a.next_multiple_of(...))`.